src/main.rs

```diff
@@ -450,110 +593,213 @@ fn get_setup(args: &[String]) -> Setup {
         list_backends();
         exit(0);
     }
 
-    let backend = audio_backend::find(backend_name).expect("Invalid backend");
+    let backend = audio_backend::find(backend_name).unwrap_or_else(|| {
+        error!(
+            "Invalid `--{}` / `-{}`: {}",
+            BACKEND,
+            BACKEND_SHORT,
+            matches.opt_str(BACKEND).unwrap_or_default()
+        );
+        list_backends();
+        exit(1);
+    });
 
     let format = matches
         .opt_str(FORMAT)
         .as_deref()
-        .map(|format| AudioFormat::from_str(format).expect("Invalid output format"))
+        .map(|format| {
+            AudioFormat::from_str(format).unwrap_or_else(|_| {
+                error!("Invalid `--{}` / `-{}`: {}", FORMAT, FORMAT_SHORT, format);
+                println!(
+                    "Valid `--{}` / `-{}` values: F64, F32, S32, S24, S24_3, S16",
+                    FORMAT, FORMAT_SHORT
+                );
+                println!("Default: {:?}", AudioFormat::default());
+                exit(1);
+            })
+        })
         .unwrap_or_default();
 
+    #[cfg(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    ))]
     let device = matches.opt_str(DEVICE);
+
+    #[cfg(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    ))]
     if device == Some("?".into()) {
         backend(device, format);
         exit(0);
     }
 
+    #[cfg(not(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    )))]
+    let device: Option<String> = None;
+
+    #[cfg(not(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    )))]
+    if matches.opt_present(DEVICE) {
+        warn!(
+            "The `--{}` / `-{}` option is not supported by the included audio backend(s), and has no effect.",
+            DEVICE, DEVICE_SHORT,
+        );
+    }
+
+    #[cfg(feature = "alsa-backend")]
     let mixer_type = matches.opt_str(MIXER_TYPE);
-    let mixer = mixer::find(mixer_type.as_deref()).expect("Invalid mixer");
+    #[cfg(not(feature = "alsa-backend"))]
+    let mixer_type: Option<String> = None;
+
+    let mixer = mixer::find(mixer_type.as_deref()).unwrap_or_else(|| {
+        error!(
+            "Invalid `--{}` / `-{}`: {}",
+            MIXER_TYPE,
+            MIXER_TYPE_SHORT,
+            matches.opt_str(MIXER_TYPE).unwrap_or_default()
+        );
+        println!(
+            "Valid `--{}` / `-{}` values: alsa, softvol",
+            MIXER_TYPE, MIXER_TYPE_SHORT
+        );
+        println!("Default: softvol");
+        exit(1);
+    });
 
     let mixer_config = {
-        let mixer_device = match matches.opt_str("mixer-card") {
-            Some(card) => {
-                warn!("--mixer-card is deprecated and will be removed in a future release.");
```

Comment:
We should document in the changelog which previously deprecated features have now been removed.

Reply:
Can do.

Reply:
Done.